File: dotnet/ArqLibrarianClassic/ArqLibrarianClassic/Library/BooksManager.cs
using System.Collections.Generic;
using System.Linq;

namespace ArqLibrarianClassic.Library
{
    public class BooksManager
    {
        private readonly BooksDao dao = null;

        public BooksManager(BooksDao dao)
        {
            this.dao = dao;
        }

        public IEnumerable<Book> FindAll()
        {
            return dao.FindAll();
        }

        public void Create(string title, string author, string isbn, string publisher, int year, string category)
        {    
            dao.Insert(new Book(title, author, isbn, publisher, year, category));
        }

        public IEnumerable<Book> FindByTitle(string title)
        {
            return dao.FindByTitle(title);
        }

        public void Rate(long id, int rating)
        {
            Book book = dao.FindById(id);

            book.Add(new SingleRating(rating));

            dao.Save(book);
        }

        public Book FindById(long id)
        {
            return dao.FindById(id);
        }

        public double ComputeRatingFor(long bookId)
        {
            var book = dao.FindById(bookId);

            if (book.Ratings.Count == 0)
            {
                return -1.0;
            }
            
            
            return book.Ratings.Average(b => b.Rating);
        }
    }
}

File: java/ArqLibraryClassic/ArqLibraryLayered/src/bnsit/ms/arq/library/BooksManager.java
package bnsit.ms.arq.library;

import java.util.Iterator;

public class BooksManager {
    private BooksDao dao;

    public BooksManager(BooksDao dao) {
        this.dao = dao;
    }

    public void create(String title, String author, String isbn, String publisher, int year, String category) {
        dao.insert(new Book(title, author, isbn, publisher, year, category));
    }

    public Iterator<Book> findAll() {
        return dao.findAll();
    }

    public Iterator<Book> findByTitle(String toSearch) {
        return dao.findByTitle(toSearch);
    }

    public Book findById(long bookId) {
        return dao.findById(bookId);
    }

    public double computeRatingFor(long id) {
        Book book = dao.findById(id);

        if (book.getRatings().size() == 0)
        {
            return -1.0;
        }

        return book.getRatings().stream()
                .mapToDouble(r -> r.getRating())
                .average().getAsDouble();

    }

    public void rate(long bookId, int rating) {
        Book book = dao.findById(bookId);

        book.add(new SingleRating(rating));

        dao.save(book);
    }
}
